### PR TITLE
Hotkeys: Ctrl+0, Ctrl+1, ... in browse window, F4 for edit etc.

### LINKED ISSUE #5347: Browse Window: Ctrl+1, Ctrl+2, ... hotkeys
**Do you want to request a *feature* or report a *bug*?**
incomplete new feature, discussion

**What is the current behavior?**
Ctrl+1, Ctrl+2, ... hotkeys do not fit for the new layout options
Some panels cannot be focused using such hotkeys.

**What is the expected behavior?**
1. Each area shall have a dedicated hotkey `Ctrl+Number`.
1. The numbering shall be intuitive and easy to remember
(as in older versions and in the `Commit` dialog: horizontal first, then vertical, i.e. row-wise)
1. The assignment should not change for different layouts. Thoughts?

Which layout shall be the base for the numbering?
grafik
This numbering would also work for
grafik

Are there more layouts that should be considered?

Especially, how shall the brandnew layout with diff info left to the graph be dealt with?
I guess users will not often switch between the left and the right position. That's why the above rule 1 is not very important. Then in this case, Ctrl+1 should focus the diff and Ctrl+2 the list of revisions.
grafik

**Environment you encounter the issue:**

- GitExtensions version: master, 2.51
- GIT version: (2.17.1.w.2-x64)
- OS version: (Windows 7)
- .NET version: (4.7)

**Did this work in previous version of GitExtensions (which)?**
2.49 because its layout could not be changed

### LINKED ISSUE #5348: Hotkey F4 to open file in editor
**Do you want to request a *feature* or report a *bug*?**
feature

**What is the current behavior?**
No hotkey is assigned/assignable to edit and to open a file in the editor.

**What is the expected behavior?**
Hotkey assignable to the context menu items following below.
Use `F4` as default which would be similar to Norton/Midnight/Total Commander's `F3` View, `F4` Edit.

Additionally in the `File Tree` and the `Commit` dialog:
Assign shortcut `Shift+F4` to context menu item `Open` and `Shift+Ctrl+F4` to `Open with...`.

grafik

grafik

grafik

grafik

**Environment you encounter the issue:**

- GitExtensions version: (master)
- GIT version: (2.17.1.w.2-x64)
- OS version: (Windows 7)
- .NET version: (4.7)

**Did this work in previous version of GitExtensions (which)?**
no, new feature

### PR DESCRIPTION
**Pull request by mstv**
Fixes #5347.
Fixes #5348.

Changes proposed in this pull request:
- adapt and extend the hotkeys for focusing different areas of the `browse window`
- add hotkeys for `edit file`, `open`, `open with...`
- sort hotkeys nearly alphabetically
 
Screenshots before and after (if PR changes UI):
- before
grafik
- after
grafik
grafik
- before/after
grafik  grafik

grafik  grafik

grafik  grafik

grafik  grafik

What did I do to test the code and ensure quality:
- tested all the hotkeys
- hit `Ctrl+Tab` and `Shift+Ctrl+Tab` to switch between Diff and FileTree
- hid/showed the `branch tree panel`
- switched the positions of the `commit info`

Has been tested on:
- GIT 2.17.1.w.2-x64
- Windows 7

### CODE REVIEW

**Comment by mstv** [comment]
CodeFactor just quibbles with the long switch statement in FormCommit.ExecuteCommand().

**Comment by RussKie** [comment]
Overall it looks good. However the shortcut names don't look correct:
image
Shouldn't `CTRL+1` be "FocusRevisionGraph", and `CTRL+2` be "FocusCommitInfo"?

I also feel "FocusFileTree" and "FocusRevisionContent" should be together (5 and 6)

**Comment by vbjay** [comment]
What's wrong with tab?  This seems to be a bit forced. It can be a
maintenance nightmare if UI changes.

**Comment by mstv** [comment]
`Tab` and `Shift+Tab` do not work as you expect in the DiffView / FileEditor. You're locked in there.

> It can be a maintenance nightmare if UI changes.

The UI has changed. Try pressing Ctrl+2 for the different positions of the commit info.

**Comment by RussKie** [comment]
Sorry Michael, but I'm not liking the inconsistency.
If the commit info tab on the right or at the bottom then, to focus it, I have to use CTRL+2, and if it is on the left CTRL+2 focuses the graph, and I have to use CTRL+1...

Let's have CTRL+1 have always focus the graph, CTRL+2 - always focus the commit info, irrespective where it is.
This will also lead to simple enum names: 
* CTRL+1 - "FocusRevisionGraph", and 
* CTRL+2 - "FocusCommitInfo".

**Comment by mstv** [comment]
OK, I'll change Ctrl+1/2. I'm dispassionate regarding which consistency aspect weighs more.

>> I also feel "FocusFileTree" and "FocusRevisionContent" should be together (4 and 5 or 5 and 6)

> I added Ctrl+6 because it's hard to jump between the list and the revision content using the keyboard.
I find the numbering not completely intuitive, too. But at least it is consistent with counting the columns (see screenshot above).
An alternative could be hitting Ctrl+3 / Ctrl+4 again to jump between the list and the revision content.

> Which one is the correct designation "RevisionContent" or "RevisionContents"?

Thoughts?

**Comment by vbjay** [comment]
Idea. Open excel with a ribon. Hit alt. See the keys UI pop up?  When ____
is held down show the same type of UI but next to each section.

**Comment by RussKie** [comment]
> I like this alternative "hit Ctrl+3/4 twice" - implementing it.

Sounds good

> Open excel with a ribon. Hit alt. 

Not easy (if possible at all) to implement in bog stock standard WinForms app

**Comment by vbjay** [comment]
:D  Didn't say it was easy.  We would need to write a ___Provider that
extends controls like the tooltip provider.

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 1]
```diff
@@ -399,7 +399,7 @@ protected override void OnLoad(EventArgs e)
             RevisionGrid.Load();
             _filterBranchHelper.InitToolStripBranchFilter();
 
-            RevisionGrid.Focus();
+            ActiveControl = RevisionGrid;
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormBrowse.cs]: makes hotkeys work right after starting GE

File: GitUI/CommandsDialogs/RevisionDiffControl.cs [hunk 2]
```diff
@@ -172,8 +172,6 @@ protected override void OnRuntimeLoad()
             DiffText.Font = AppSettings.FixedWidthFont;
             ReloadHotkeys();
 
-            GotFocus += (s, e1) => DiffFiles.Focus();
```
Comment (mstv) [inline GitUI/CommandsDialogs/RevisionDiffControl.cs]: This event handler was not called at all.

File: GitUI/CommandsDialogs/RevisionFileTreeControl.cs [hunk 3]
```diff
@@ -288,8 +288,6 @@ protected override void OnRuntimeLoad()
 
             ReloadHotkeys();
 
-            GotFocus += (s, e1) => tvGitTree.Focus();
```
Comment (mstv) [inline GitUI/CommandsDialogs/RevisionFileTreeControl.cs]: This event handler was not called at all.

File: GitUI/Hotkey/HotkeySettingsManager.cs [hunk 4]
```diff
@@ -243,49 +243,69 @@ public static HotkeySettings[] CreateDefaultSettings()
...
+                    Hk(RevisionGridControl.Commands.NextQuickSearch, Keys.Alt | Keys.Down),
+                    Hk(RevisionGridControl.Commands.PrevQuickSearch, Keys.Alt | Keys.Up),
                     Hk(RevisionGridControl.Commands.RevisionFilter, Keys.Control | Keys.F),
+                    Hk(RevisionGridControl.Commands.SelectCurrentRevision, Keys.Control | Keys.Shift | Keys.C),
+                    Hk(RevisionGridControl.Commands.SelectAsBaseToCompare, Keys.Control | Keys.L),
+                    Hk(RevisionGridControl.Commands.ShowAllBranches, Keys.Control | Keys.Shift | Keys.A),
+                    Hk(RevisionGridControl.Commands.ShowCurrentBranchOnly, Keys.Control | Keys.Shift | Keys.U),
+                    Hk(RevisionGridControl.Commands.ShowFilteredBranches, Keys.Control | Keys.Shift | Keys.T),
+                    Hk(RevisionGridControl.Commands.ShowFirstParent, Keys.Control | Keys.Shift | Keys.S),
+                    Hk(RevisionGridControl.Commands.ShowRemoteBranches, Keys.Control | Keys.Shift | Keys.R),
```
Comment (RussKie) [inline GitUI/Hotkey/HotkeySettingsManager.cs]: Love the A-Z order 👍

**Comment by RussKie** [comment]
I like the idea of using the same shortcut to move between tabs, nice work 👍

**Comment by RussKie** [comment]
I am happy to merge it, please rebase on top of the latest master.
Thanks

**Comment by mstv** [comment]
rebase done

**Comment by mstv** [comment]
Ctrl+F has not been touched by this PR at all. Yes, it could be handled in a similar way as F3, F4 etc.
But this should be discussed in a new dedicated issue because:
- This PR deals with the adaptation/completion of hotkeys for focusing and new hotkeys for Edit/Open File. It contains already enough changes for one PR in my opinion.
- The hotkey management might need a redesign towards a kind of inheritance/delegation because more hotkeys would need to be configured twice -- for Browse and for BrowseDiff/RevisionFileTree.

**Comment by gerhardol** [comment]
Typo in my prev post should be Ctrl-F3, not Ctrl-F#

**Comment by RussKie** [comment]
We can always do further tweak later :)

Thank you all

**Pull request merged**